src/runtime.rs

```diff
@@ -0,0 +1,154 @@
+//! Tokio runtime
+
+use reactor::{self, Reactor};
+use reactor::background::{self, Background};
+
+use tokio_executor;
+use tokio_threadpool::{self as threadpool, ThreadPool};
+use futures::Poll;
+use futures::future::{Future, Join};
+
+use std::io;
+
+/// Handle to the Tokio runtime.
+///
+/// The Tokio runtime includes a reactor as well as an executor for running
+/// tasks.
+#[derive(Debug)]
+pub struct Runtime {
+    inner: Option<Inner>,
+}
+
+/// Error returned from `with_context`.
+#[derive(Debug)]
+pub struct EnterError(());
+
+/// A future that resolves when the Tokio `Runtime` is shut down.
+#[derive(Debug)]
+pub struct Shutdown {
+    inner: Join<background::Shutdown, threadpool::Shutdown>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    /// Reactor running on a background thread.
+    reactor: Background,
+
+    /// Task execution pool.
+    pool: ThreadPool,
+}
+
+// ===== impl Runtime =====
+
+/// Start the Tokio runtime.
+pub fn run<F: FnOnce() -> R, R>(f: F) -> R {
+    let mut runtime = Runtime::new().unwrap();
+    let ret = runtime.with_context(f).unwrap();
+
+    runtime.shutdown().wait().unwrap();
+
+    ret
+}
+
+/// Start the Tokio runtime and spawn the provided future.
+pub fn run_seeded<F>(f: F)
+where F: Future<Item = (), Error = ()> + Send + 'static,
+{
+    let mut runtime = Runtime::new().unwrap();
+    runtime.spawn(f);
+    runtime.shutdown().wait().unwrap();
+}
+
+impl Runtime {
+    /// Create a new runtime
+    pub fn new() -> io::Result<Self> {
+        // Spawn a reactor on a background thread.
+        let reactor = Reactor::new()?.background()?;
+
+        // Get a handle to the reactor.
+        let handle = reactor.handle().clone();
+
+        let pool = threadpool::Builder::new()
```

Comment:
we probably want a way to configure the upper bounds of number of threads this threadpool may run?